modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -328,106 +344,207 @@ def find_esc3_vuln_cert_templates
-      print_error("Unable to retrieve the username using ldapwhoami, ESC4 detection can't continue")
-      return
-    end
+    return if esc_entries.empty?
 
-    sam_account_name = whoami_response.split('\\')[1]
-    user_raw_filter = "(sAMAccountName=#{sam_account_name})"
-    attributes = ['DN', 'objectSID', 'objectClass', 'primarygroupID']
-    our_account = query_ldap_server(user_raw_filter, attributes)&.first
-    if our_account.nil?
-      print_warning("Unable to determine the User SID for #{sam_account_name}, ESC4 detection can't continue")
-      return
+    current_user = adds_get_current_user(@ldap)[:samaccountname].first
+    esc_entries.each do |entry|
+      certificate_symbol = entry[:cn][0].to_sym
+      next if @certificate_details[certificate_symbol][:enroll_sids].empty?
+
+      if adds_obj_grants_permissions?(@ldap, entry, SecurityDescriptorMatcher::Allow.any(%i[WP]))
+        @certificate_details[certificate_symbol][:techniques] << 'ESC4'
+        @certificate_details[certificate_symbol][:notes] << "ESC4: The account: #{current_user} has edit permissions over the template #{certificate_symbol}."
+      end
     end
+  end
 
-    user_sid = map_sids_to_names([Rex::Proto::MsDtyp::MsDtypSid.read(our_account[:objectsid].first).value]).first
-    domain_sid = user_sid.value.to_s.rpartition('-').first
-    user_groups = []
+  def parse_registry_output(output, property_name)
+    return nil if output.stderr.present?
 
-    if our_account[:primarygroupID]
-      user_groups << "#{domain_sid}-#{our_account[:primarygroupID]&.first}"
+    stdout = output.stdout if output.stdout.present?
+    return nil unless stdout
+
+    line_with_property = stdout.lines.find { |line| line.strip.match(/^#{Regexp.escape(property_name)}\s*:/) }
+    return nil unless line_with_property
+
+    line_with_property.split(':', 2).last&.strip
+  end
+
+  def run_registry_command(shell, path, property_name, dynamic_value = nil)
+    full_path = dynamic_value ? "#{path}\\#{dynamic_value}" : path
+    command = "Get-ItemProperty -Path '#{full_path}' -Name #{property_name}"
+    output = shell.run(command)
+    value = parse_registry_output(output, property_name)
+    if value.nil?
+      print_error("Registry property '#{property_name}' not found at path '#{full_path}'.")
     end
+    value
+  end
+
+  def enum_registry_values
+    @registry_values ||= {}
 
-    # Authenticated Users includes all users and computers with identities that have been authenticated.
-    # Authenticated Users doesn't include Guest even if the Guest account has a password.
-    unless sam_account_name == 'Guest'
-      user_groups << Rex::Proto::Secauthz::WellKnownSids::SECURITY_AUTHENTICATED_USER_SID
+    endpoint = "http://#{datastore['RHOST']}:5985/wsman"
+    user = datastore['LDAPUsername']
+    pass = datastore['LDAPPassword']
+
+    conn = WinRM::Connection.new(
+      endpoint: endpoint,
+      user: user,
+      password: pass,
+      transport: :negotiate
+    )
```

Comment:
LDAP authentication is kinda complex, there's a bunch of ways the user can specify their auth info. I'd suggest pulling the username and domain dynamically to pass back to WinRM. This would account for the edge case where `LDAP::Auth` is `plaintext` and the username is in the format `name@domain` without you needing to check the auth type and parse the name.

This code here will also fail to authenticate when Kerberos or Schannel authentication is used. In that case, I'd add a check to the `#validate` method to raise an option validation error on `RUN_REGISTRY_CHECKS` being incompatible with `LDAP::Auth`.

```suggestion
    domain = adds_get_domain_info(@ldap)[:dns_name]
    user = adds_get_current_user(@ldap)[:sAMAccountName].first.to_s
    pass = datastore['LDAPPassword']

    conn = WinRM::Connection.new(
      endpoint: endpoint,
      domain: domain,
      user: user,
      password: pass,
      transport: :negotiate
    )
```